README.md
[![Build Status](https://travis-ci.org/Yelp/kafka-utils.svg?branch=master)](https://travis-ci.org/Yelp/kafka-utils)

# Kafka-Utils

A suite of python tools to interact and manage Apache Kafka clusters.
Kafka-Utils runs on python2.7 and python3.

## Configuration

Kafka-Utils reads cluster configuration needed to access Kafka clusters from yaml files. Each cluster is identified by *type* and *name*.
Multiple clusters of the same type should be listed in the same `type.yaml` file.
The yaml files are read from `$KAFKA_DISCOVERY_DIR`, `$HOME/.kafka_discovery` and `/nail/srv/configs/kafka_discovery_configs`, the former overrides the latter.
If the `/nail/srv/configs/kafka_discovery_configs` directory does not exist then the `/etc/kafka_discovery` directory is used instead.


Sample configuration for `sample_type` cluster at `/nail/srv/configs/kafka_discovery_configs/sample_type.yaml`

```yaml
---
  clusters:
    cluster-1:
      broker_list:
        - "cluster-elb-1:9092"
      zookeeper: "11.11.11.111:2181,11.11.11.112:2181,11.11.11.113:2181/kafka-1"
    cluster-2:
      broker_list:
        - "cluster-elb-2:9092"
      zookeeper: "11.11.11.211:2181,11.11.11.212:2181,11.11.11.213:2181/kafka-2"
  local_config:
    cluster: cluster-1
```

## Install

From PyPI:
```shell
    $ pip install kafka-utils
```


## Kafka-Utils command-line interface

### List all clusters

```shell
    $ kafka-utils
    cluster-type sample_type:
        cluster-name: cluster-1
        broker-list: cluster-elb-1:9092
        zookeeper: 11.11.11.111:2181,11.11.11.112:2181,11.11.11.113:2181/kafka-1
        cluster-name: cluster-2
        broker-list: cluster-elb-2:9092
        zookeeper: 11.11.11.211:2181,11.11.11.212:2181,11.11.11.213:2181/kafka-2
```

### Get consumer offsets

```shell
    $ kafka-consumer-manager --cluster-type sample_type offset_get sample_consumer
```

### Get consumer watermarks

```shell
    $ kafka-consumer-manager --cluster-type sample_type get_topic_watermark sample.topic

```


### Rebalance cluster cluster1 of type sample_cluster

```shell
    $ kafka-cluster-manager --cluster-type sample_type --cluster-name cluster1
    --apply rebalance --brokers --leaders --max-partition-movements 10
    --max-leader-changes 15
```

### Rolling-restart a cluster

```shell
    $ kafka-rolling-restart --cluster-type sample_type
```

### Check in-sync replicas

```shell
    $ kafka-check --cluster-type sample_type min_isr
```

### Check number of unavailable replicas

```shell
    $ kafka-check --cluster-type sample_type replica_unavailability
```

## Documentation

Read the documentation at [Read the Docs](http://kafka-utils.readthedocs.io/en/latest/).

## License

Kafka-Utils is licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0

## Contributing

Everyone is encouraged to contribute to Kafka-Utils by forking the
[Github repository](http://github.com/Yelp/kafka-utils) and making a pull request or opening an issue.


CHANGELOG.rst
3.9.0 (November 13, 2020)
----------------------------
* Change the default Kafka discovery config directory back to /etc/kafka_discovery

3.8.2 (November 9, 2020)
----------------------------
* Conditionally choose between /nail/srv/configs/kafka_discovery_configs and /etc/kafka_discovery for default directory to read Kafka discovery files depending on if the former exists

3.8.1 (November 9, 2020)
----------------------------
* Pin version of cryptography to 3.0

3.8.0 (November 6, 2020)
----------------------------
* Change the default directory to read kafka_discovery files from /etc/kafka_discovery to /nail/srv/configs/kafka_discovery_configs

3.7.1 (August 17, 2020)
----------------------------
* Fix missing rf_mismatch arg in process_assignment call

3.7.0 (June 25, 2020)
----------------------------
* Allow dynamic_import to find the child-most class

3.6.0 (April 2, 2020)
----------------------------
* Allow kafka-rolling-restart script to be executed in clusters with SSL or SASL enabled (thanks dinegri!)

3.5.0 (March 10, 2020)
----------------------------
* Add --use-admin-client option to kafka-consumer-manager commands increased faster performance.

3.4.0 (March 4, 2020)
----------------------------
* Allow set_replication_factor command to work with topics that have mismatched replication
  factors in their partitions
* Add new preferred replica election command to generate json files for kafka-preferred-replica-election

3.3.0 (February 26, 2020)
----------------------------
* Allow for removing brokers in kafka-cluster-manager replace command
* Support consumer offset message value schema v3 (thanks mborst!)

3.2.0 (January 3, 2020)
----------------------------
* Add --active-controller-for-last flag to kafka-rolling-restart

3.1.0 (November 5, 2019)
----------------------------
* Use v1.4.7 of kafka-python to take advantage of performance upgrades
* Fix Travis build by unpinning Docker version

3.0.3 (September 18, 2019)
----------------------------
* Fix git tag

3.0.2 (September 18, 2019)
----------------------------
* Relax pytz pinned version in setup.py

3.0.1 (July 19, 2019)
----------------------------
* Fix configuration writing in kafka-manual-throttle

3.0.0 (July 16, 2019)
----------------------------
* Remove support for python3.4

2.3.1 (July 08, 2019)
----------------------------
* Fix iteration bug in _extract_actions_unique_topics with --max-movement-size

2.3.0 (May 08, 2019)
----------------------------
* Implement --head option for kafka-check command

2.2.3 (May 03, 2019)
----------------------------
* Fix bug with --max-movement-size default value causing stuck decommissions

2.2.2 (April 30, 2019)
----------------------------
* Relax requirements for PyYAML and pytz

2.2.1 (April 26, 2019)
----------------------------
* Use openjdk instead of oracle-java8 in acceptance tests
* Migrate to ubuntu 16.04 in acceptance tests
* Update team name and email

2.2.0 (April 8, 2019)
----------------------------
* Add new --force-progress flag in kafka-cluster-manager decommission
* Fix bug with moving 0-size partitions in decommission
* Fix bug with genetic balancer using size in decommission when --max-movement-size not specified

2.1.0 (April 3, 2019)
----------------------------
* Add new --max-movement-size and --auto-max-movement-size in kafka-cluster-manager decommission
* Upgrade dependency requirements to newer versions
* Fix bug with KafkaGroupReader acceptance test

2.0.0 (January 29, 2019)
----------------------------
* Deprecate zookeeper offset storage in kafka-consumer-manager

1.8.0 (January 3, 2019)
----------------------------
* Implement kafka-check replication_factor command

1.7.5 (December 3, 2018)
----------------------------
* Fix imports, pytest version and build failure due to flake8

1.7.4 (Novemeber 26, 2018)
----------------------------
* Add change to display generated reassignment plan even on failure of validation

1.7.3 (October 11, 2018)
----------------------------
* Fix bug in offset_get command that showed topics as unsubscribed if any
  partition had an offset of 0 committed

1.7.2 (October 3, 2018)
----------------------------
* Upgrade paramiko in kafka-utils (paramiko < 2.5.0)

1.7.1 (September 7, 2018)
----------------------------
* Fix bug in unsubscribe_topics command where all subscribed topics were being
  displayed as subject to change regardless of specified --topics

1.7.0 (September 6, 2018)
----------------------------
* Add new commands offset_set_timestamp and offsets_for_timestamp

1.6.8 (August 15, 2018)
----------------------------
* Add verbosity option for kafka-consumer-manager

1.6.7 (August 15, 2018)
----------------------------
* Bump setuptools version.

1.6.6 (August 14, 2018)
----------------------------
* Remove cryptography dependency

1.6.5 (July 16th, 2018)
----------------------------
* Add tox extension
* Upgrade kafka-python version to 1.4.2

1.6.4 (June 13th, 2018)
----------------------------
* Improve performance of broker decommission process

1.6.3 (June 7th, 2018)
----------------------------
* kafka-check offline_partitions does not throw an exception for empty clusters anymore
* Add --broker-ids option to kafka-rolling-restart

1.6.2 (June 5th, 2018)
----------------------------
* Fix offset_get and delete_group when no offsets folder under zk consumers node

1.6.1 (May 31st, 2018)
----------------------------
* Add NoNodeError to get_brokers & get_topics
* Catch Exceptions for commands under empty cluster exception

1.6.0 (May 29th, 2018)
----------------------------
* Set exactly topic matching as default for get_topic_watermark command
* Add -r/--regex option for get_topic_watermark command for regex search

1.5.1 (May 11th, 2018)
----------------------------
* Improve handling of missing topic in getting topic-specific configuration

1.5.0 (April 4th, 2018)
----------------------------
* Add --topics option to kafka_consumer_manager unsubscribe_topics

1.4.2 (March 20th, 2018)

kafka_tools/util/config.py
# -*- coding: utf-8 -*-
# Copyright 2016 Yelp Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import glob
import logging
import os
from collections import namedtuple

import yaml

from kafka_tools.util.error import ConfigurationError
from kafka_tools.util.error import InvalidConfigurationError
from kafka_tools.util.error import MissingConfigurationError


DEFAULT_KAFKA_TOPOLOGY_BASE_PATH = '/etc/kafka_discovery'
HOME_OVERRIDE = '.kafka_discovery'


class ClusterConfig(
    namedtuple(
        'ClusterConfig',
        ['type', 'name', 'broker_list', 'zookeeper'],
    ),
):
    """Cluster configuration.
    :param name: cluster name
    :param broker_list: list of kafka brokers
    :param zookeeper: zookeeper connection string
    """

    def __ne__(self, other):
        return self.__hash__() != other.__hash__()

    def __eq__(self, other):
        return self.__hash__() == other.__hash__()

    def __hash__(self):
        if isinstance(self.broker_list, list):
            broker_list = self.broker_list
        else:
            broker_list = self.broker_list.split(',')
        zk_list = self.zookeeper.split(',')
        return hash((
            self.type,
            self.name,
            ",".join(sorted(filter(None, broker_list))),
            ",".join(sorted(filter(None, zk_list)))
        ))


def load_yaml_config(config_path):
    with open(config_path, 'r') as config_file:
        return yaml.safe_load(config_file)


class TopologyConfiguration(object):
    """Topology configuration for a kafka cluster.

    Read a cluster_type.yaml from the kafka_topology_path.
    Example config file:
    .. code-block:: yaml

       clusters:
         cluster1:
             broker_list:
               - "broker1:9092"
               - "broker2:9092"
             zookeeper: "zookeeper1:2181/mykafka"
         cluster2:
             broker_list:
               - "broker3:9092"
               - "broker4:9092"
             zookeeper: "zookeeper2:2181/mykafka"
       local_config:
         cluster: cluster1


    :param cluster_type: kafka cluster type.
    :type cluster_type: string
    :param kafka_topology_path: path of the directory containing
        the kafka topology.yaml config
    :type kafka_topology_path: string
    """

    def __init__(
        self,
        cluster_type,
        kafka_topology_path=DEFAULT_KAFKA_TOPOLOGY_BASE_PATH
    ):
        self.kafka_topology_path = kafka_topology_path
        self.cluster_type = cluster_type
        self.log = logging.getLogger(self.__class__.__name__)
        self.clusters = None
        self.local_config = None
        self.load_topology_config()

    def __eq__(self, other):
        if all([
            self.cluster_type == other.cluster_type,
            self.clusters == other.clusters,
            self.local_config == other.local_config,
        ]):
            return True
        return False

    def __ne__(self, other):
        return not self.__eq__(other)

    def load_topology_config(self):
        """Load the topology configuration"""
        config_path = os.path.join(
            self.kafka_topology_path,
            '{id}.yaml'.format(id=self.cluster_type),
        )
        self.log.debug("Loading configuration from %s", config_path)
        if os.path.isfile(config_path):
            topology_config = load_yaml_config(config_path)
        else:
            raise MissingConfigurationError(
                "Topology configuration {0} for cluster {1} "
                "does not exist".format(
                    config_path,
                    self.cluster_type,
                )
            )
        self.log.debug("Topology configuration %s", topology_config)
        try:
            self.clusters = topology_config['clusters']
            self.local_config = topology_config['local_config']
        except KeyError:
            self.log.exception("Invalid topology file")
            raise InvalidConfigurationError("Invalid topology file {0}".format(
                config_path))

    def get_all_clusters(self):
        return [
            ClusterConfig(
                type=self.cluster_type,
                name=name,
                broker_list=cluster['broker_list'],
                zookeeper=cluster['zookeeper'],
            )
            for name, cluster in self.clusters.iteritems()
        ]

    def get_cluster_by_name(self, name):
        if name in self.clusters:
            cluster = self.clusters[name]
            return ClusterConfig(
                type=self.cluster_type,
                name=name,
                broker_list=cluster['broker_list'],
                zookeeper=cluster['zookeeper'],
            )
        raise ConfigurationError("No cluster with name: {0}".format(name))

    def get_local_cluster(self):
        try:
            if self.local_config:
                local_cluster = self.clusters[self.local_config['cluster']]
                return ClusterConfig(
                    type=self.cluster_type,
                    name=self.local_config['cluster'],
                    broker_list=local_cluster['broker_list'],
                    zookeeper=local_cluster['zookeeper'])
        except KeyError:
            self.log.exception("Invalid topology file")
            raise InvalidConfigurationError("Invalid topology file")

    def __repr__(self):
        return ("TopologyConfig: cluster_type {0}, clusters: {1},"
                "local_config {2}".format(
                    self.cluster_type,
                    self.clusters,
                    self.local_config
                ))


def get_conf_dirs():
    config_dirs = []
    if "KAFKA_DISCOVERY_DIR" in os.environ and os.environ["KAFKA_DISCOVERY_DIR"]:
        config_dirs.append(os.environ["KAFKA_DISCOVERY_DIR"])
    if os.environ["HOME"]:
        home_config = os.path.join(
            os.path.abspath(os.environ['HOME']),
            HOME_OVERRIDE,
        )
        if os.path.isdir(home_config):
            config_dirs.append(home_config)
    config_dirs.append(DEFAULT_KAFKA_TOPOLOGY_BASE_PATH)
    return config_dirs


def get_cluster_config(
    cluster_type,
    cluster_name=None,
    kafka_topology_base_path=None,
):
    """Return the cluster configuration.
    Use the local cluster if cluster_name is not specified.

    :param cluster_type: the type of the cluster
    :type cluster_type: string
    :param cluster_name: the name of the cluster
    :type cluster_name: string
    :param kafka_topology_base_path: base path to look for <cluster_type>.yaml
    :type cluster_name: string
    :returns: the cluster
    :rtype: ClusterConfig
    """
    if not kafka_topology_base_path:
        config_dirs = get_conf_dirs()
    else:
        config_dirs = [kafka_topology_base_path]

    topology = None
    for config_dir in config_dirs:
        try:
            topology = TopologyConfiguration(
                cluster_type,
                config_dir,
            )
        except MissingConfigurationError:
            pass
    if not topology:
        raise MissingConfigurationError(
            "No available configuration for type {0}".format(cluster_type),
        )

    if cluster_name:
        return topology.get_cluster_by_name(cluster_name)
    else:
        return topology.get_local_cluster()


def iter_configurations(kafka_topology_base_path=None):
    """Cluster topology iterator.
    Iterate over all the topologies available in config.
    """
    if not kafka_topology_base_path:
        config_dirs = get_conf_dirs()
    else:
        config_dirs = [kafka_topology_base_path]

    types = set()
    for config_dir in config_dirs:
        new_types = filter(
            lambda x: x not in types,
            map(
                lambda x: os.path.basename(x)[:-5],
                glob.glob('{0}/*.yaml'.format(config_dir)),
            )
        )
        for cluster_type in new_types:
            try:
                topology = TopologyConfiguration(
                    cluster_type,
                    config_dir,
                )
            except ConfigurationError:
                continue
            types.add(cluster_type)
            yield topology
